craft_application/services/service_factory.py

```diff
@@ -35,22 +62,24 @@ class ServiceFactory:
     and possibly have its existing service classes overridden.
     """
 
-    app: AppMetadata
-
-    PackageClass: type[services.PackageService]
-    LifecycleClass: type[services.LifecycleService] = services.LifecycleService
-    ProviderClass: type[services.ProviderService] = services.ProviderService
-    RemoteBuildClass: type[services.RemoteBuildService] = services.RemoteBuildService
-    RequestClass: type[services.RequestService] = services.RequestService
-    ConfigClass: type[services.ConfigService] = services.ConfigService
-    FetchClass: type[services.FetchService] = services.FetchService
-    InitClass: type[services.InitService] = services.InitService
+    _service_classes: ClassVar[
+        dict[str, tuple[str, str] | type[services.AppService]]
+    ] = {}
 
+    # These exist so that child ServiceFactory classes can use them.
+    app: AppMetadata
+    PackageClass: type[services.PackageService] = None  # type: ignore[assignment]
+    LifecycleClass: type[services.LifecycleService] = None  # type: ignore[assignment]
```

Comment:
I haven't looked too much at the tests, but is there a test for the case where the ServiceFactory subclass overrides one of these? E.g., a test that overrides `LifecycleClass` and then checks that factory.lifecycle is of the correct class.

Since this is to be backwards-compatible I think that's our main compatibility usecase

Reply:
That's fair - I've done IRL testing with Charmcraft but I didn't actually add tests here for it. Thanks!